File: android/app/src/main/java/dev/veeso/biangbianghanzi/services/PinyinTransliterator.kt
package dev.veeso.biangbianghanzi.services

import dev.veeso.biangbiangui.protocols.Transliterator
import net.sourceforge.pinyin4j.PinyinHelper
import net.sourceforge.pinyin4j.format.HanyuPinyinCaseType
import net.sourceforge.pinyin4j.format.HanyuPinyinOutputFormat
import net.sourceforge.pinyin4j.format.HanyuPinyinToneType
import net.sourceforge.pinyin4j.format.HanyuPinyinVCharType

/**
 * Mandarin romanisation. The library's
 * [dev.veeso.biangbiangui.services.TextProcessingEngine] owns span
 * detection, passthrough and spacing; this only romanises one isolated
 * script span via `pinyin4j` (faithfully the former
 * `TextProcessor.hanziToPinyin`). `pinyin4j` is an app-only dependency,
 * never referenced by the library.
 */
class PinyinTransliterator : Transliterator {

    private val format: HanyuPinyinOutputFormat = HanyuPinyinOutputFormat().apply {
        caseType = HanyuPinyinCaseType.LOWERCASE
        toneType = HanyuPinyinToneType.WITH_TONE_MARK
        vCharType = HanyuPinyinVCharType.WITH_U_UNICODE
    }

    override fun transliterate(scriptSpan: String): String =
        PinyinHelper.toHanYuPinyinString(scriptSpan, format, " ", true)
            .trim()
}

File: ios/BiangBiang Hanzi/Services/PinyinTransliterator.swift
//
//  PinyinTransliterator.swift
//  BiangBiang Hanzi
//
//  Mandarin romanisation. The library's `TextProcessingEngine` owns span
//  detection, passthrough and spacing; this only romanises one isolated
//  script span via a single `CFStringTransform` to Latin (faithfully the
//  former `TextProcessor.hanziToPinyin`).
//

import BiangBiangUI
import Foundation

struct PinyinTransliterator: Transliterator {
    func transliterate(_ scriptSpan: String) -> String {
        let mutString = NSMutableString(string: scriptSpan) as CFMutableString
        CFStringTransform(mutString, nil, kCFStringTransformToLatin, false)
        return mutString as String
    }
}
